handle irqbalance service

pperiyasamy opened the pull request:
This would manipulate irqbalance service if it's already running on the system to exclude assigned pod CPUs from IRQ balancing.


Signed-off-by: Periyasamy Palanisamy <periyasamy.palanisamy@est.tech>

<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/master/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

> /kind api-change
> /kind bug
> /kind ci
> /kind cleanup
> /kind dependency-change
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
> /kind other

#### What this PR does / why we need it:
The [crio ](https://github.com/cri-o/cri-o/pull/4022) currently runs `irqbalance --oneshot` command to  exclude pod cpus from handling interrupts, but this may not be a right choice if irqbalance service is already running on the system. This PR aims to address that concern by restarting the service after updating irqbalance config file with required settings for IRQBALANCE_BANNED_CPUS parameter.

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->
None

#### Special notes for your reviewer:
This change is not yet tested, after a initial round of reviews would go for testing out this change.
#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
When using high performance hooks, CRI-O now restarts the irqbalance service after updating the irqbalance config file. rather than calling `irqbalance --oneshot`. A new config value `irqbalance_config_file` has been introduced to configure the file to update with `IRQBALANCE_BANNED_CPUS` settings. The default of this config value is `/etc/sysconfig/irqbalance`, but must be set to `/etc/default/irqbalance` for Ubuntu-like distributions.
```

pperiyasamy commented:
@JanScheurich

TomSweeneyRedHat commented:
@pperiyasamy looks like you've some lint issues.

TomSweeneyRedHat commented:
Any tests to add or at least to modify?

pperiyasamy commented:
@TomSweeneyRedHat Thanks, golint issues are fixed now, I'm waiting for the initial feedback on the changes, if that is finalised, then I'm planning test this on my local setup and add/modify test cases.

haircommander commented:
@jianzzha @cynepco3hahue PTAL

haircommander commented on internal/runtimehandlerhooks/high_performance_hooks.go:
```diff
@@ -33,11 +33,15 @@ const (
 // HighPerformanceHooks used to run additional hooks that will configure a system for the latency sensitive workloads
-type HighPerformanceHooks struct{}
+type HighPerformanceHooks struct {
+	irqBalanceConfigFile string
```
why have this file as a member of the Hooks, when the other file locations are hardcoded?

pperiyasamy replied:
@haircommander This file location is platform dependent. For example, in Ubuntu, the config file is /etc/default/irqbalance. Hence making it to configure from a RuntimeConfig property. Can you check it now ?

jianzzha commented:
> @jianzzha @cynepco3hahue PTAL

@haircommander  I will get someone to validate this in centos

haircommander commented:
/ok-to-test

haircommander commented:
by the way, thanks for taking this on @pperiyasamy, and thanks @jianzzha for the testing help !

jianzzha commented on internal/runtimehandlerhooks/high_performance_hooks.go:
```diff
@@ -311,3 +320,35 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
 }
+
+// RestoreIrqBalanceConfig restores irqbalance service with original banned cpu mask settings
+func RestoreIrqBalanceConfig(irqBalanceConfigFile, irqBannedCPUConfigFile string) error {
```
Thanks for adding this restore function. Can we check if the irqbalance service is enabled before restart it?

pperiyasamy replied:
sure, done.

jianzzha replied:
looks good to me

jianzzha commented:
@haircommander  integration test on bare metal (centos k8s) works fine; some minor comments posted inline
@cynepco3hahue  on the openshift side - I see no extra work to adopt this
@pperiyasamy can you take a look of the inline comments

pperiyasamy commented on internal/runtimehandlerhooks/high_performance_hooks.go:
```diff
@@ -240,16 +245,20 @@ func setIRQLoadBalancing(c *oci.Container, enable bool, irqSmpAffinityFile strin
-		logrus.Warnf("irqbalance binary not found: %v", err)
-		return nil
+
+	if !fileExists(irqBalanceConfigFile) {
```
yes, good point! added is-enabled check now.

pperiyasamy commented on internal/runtimehandlerhooks/utils.go:
```diff
@@ -121,3 +125,68 @@ func UpdateIRQSmpAffinityMask(cpus, current string, set bool) (cpuMask, bannedCP
+	lines := strings.Split(string(input), "\n")
+	found := false
+	for i, line := range lines {
+		if strings.Contains(line, irqBalanceBannedCpus+"=") {
```
correct, now changed to use HasPrefix method, i guess that should be ok.

haircommander commented on docs/crio.conf.5.md:
```diff
@@ -125,6 +125,9 @@ the container runtime configuration.
 **apparmor_profile**=""
   Used to change the name of the default AppArmor profile of CRI-O. The default profile name is "crio-default".
 
+**irqbalance_config_file**="/etc/sysconfig/irqbalance"
```
can you add a blurb about where this file should be for which distro (only for the most popular: ubuntu/centos, maybe debian/rhel if they differ at all)

pperiyasamy replied:
Yes! done, I was able to confirm the file locations only in centos/suse and ubuntu platforms. so added it only for those cases.

haircommander commented on internal/runtimehandlerhooks/high_performance_hooks_test.go:
```diff
@@ -107,10 +107,11 @@ var _ = Describe("high_performance_hooks", func() {
 	})
 
-	Describe("setIRQLoadBalancing", func() {
+	Describe("setIRQLoadBalancing1", func() {
```
can these have more informative names?

pperiyasamy replied:
done.

pperiyasamy commented:
> @pperiyasamy: The following tests **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> ci/openshift-jenkins/integration_crun	[e14e386](https://github.com/cri-o/cri-o/commit/e14e3868d0b2489d196aa88a9a8a8ce126daa423)	[link](https://prow.ci.openshift.org/view/gcs/origin-federated-results/pr-logs/pull/cri-o_cri-o/4441/test_pull_request_crio_integration_crun_fedora/13453/)	`/test integration_crun`
> ci/prow/e2e-gcp	[e14e386](https://github.com/cri-o/cri-o/commit/e14e3868d0b2489d196aa88a9a8a8ce126daa423)	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/cri-o_cri-o/4441/pull-ci-cri-o-cri-o-master-e2e-gcp/1347180642726580224)	`/test e2e-gcp`
> ci/prow/e2e-agnostic	[e14e386](https://github.com/cri-o/cri-o/commit/e14e3868d0b2489d196aa88a9a8a8ce126daa423)	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/cri-o_cri-o/4441/pull-ci-cri-o-cri-o-master-e2e-agnostic/1347180642676248576)	`/test e2e-agnostic`
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=cri-o&repo=cri-o&pr=4441). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:pperiyasamy).

@haircommander while addressing your review comments, any idea while these checks are failing ? Looks it's not related to this PR.

haircommander commented:
the ci/prow/e2e* are generally flaky and I believe integration_crun was a provisioning flake
/retest

haircommander commented on docs/crio.conf.5.md:
```diff
@@ -125,6 +125,10 @@ the container runtime configuration.
 
+**irqbalance_config_file**="/etc/sysconfig/irqbalance"
+  Used to change irqbalance service config file which is used by the crio runtime.
+  For centos/suse, this file is located at /etc/sysconfig/irqbalance. For ubuntu, this file is located at /etc/default/irqbalance.
```
nit, I believe it should be CentOS/SUSE and Ubuntu

haircommander commented on docs/crio.conf.5.md:
```diff
@@ -125,6 +125,10 @@ the container runtime configuration.
   Used to change the name of the default AppArmor profile of CRI-O. The default profile name is "crio-default".
 
+**irqbalance_config_file**="/etc/sysconfig/irqbalance"
+  Used to change irqbalance service config file which is used by the crio runtime.
```
`used by CRI-O.`

haircommander commented on internal/criocli/criocli.go:
```diff
@@ -520,6 +523,11 @@ func getCrioFlags(defConf *libconfig.Config) []cli.Flag {
 		},
+		&cli.StringFlag{
+			Name:  "irqbalance-config-file",
+			Usage: "The irqbalance service config file which is used by the crio runtime.",
```
`CRI-O` rather than `the crio runtime`

pperiyasamy replied:
done

haircommander commented:
sorry @pperiyasamy a couple more nits

cynepco3hahue commented on internal/runtimehandlerhooks/utils.go:
```diff
@@ -121,3 +125,74 @@ func UpdateIRQSmpAffinityMask(cpus, current string, set bool) (cpuMask, bannedCP
+	for i, line := range lines {
+		if strings.HasPrefix(line, irqBalanceBannedCpus+"=") {
+			lines[i] = irqBalanceBannedCpus + "=" + "\"" + newIRQBalanceSetting + "\""
+			found = true
```
I think you can break after you find the needed line

pperiyasamy replied:
yes, done

cynepco3hahue commented on internal/runtimehandlerhooks/utils.go:
```diff
@@ -121,3 +125,74 @@ func UpdateIRQSmpAffinityMask(cpus, current string, set bool) (cpuMask, bannedCP
+	return "", nil
+}
+
+func trimQuotes(s string) string {
```
does not `strings.Trim("\"")` work?

pperiyasamy replied:
yes, it would work, thanks.

cynepco3hahue commented on internal/runtimehandlerhooks/utils.go:
```diff
@@ -121,3 +125,74 @@ func UpdateIRQSmpAffinityMask(cpus, current string, set bool) (cpuMask, bannedCP
+	return !info.IsDir()
+}
+
+func isServiceEnabled(serviceName string) bool {
```
Can we assume that `systemctl` will always be responsible to manage services under the operating system? If yes, can we use `systemctl` to restart the service instead of `service` command?

pperiyasamy commented on internal/runtimehandlerhooks/high_performance_hooks.go:
```diff
@@ -311,3 +333,43 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
 }
+
+// RestoreIrqBalanceConfig restores irqbalance service with original banned cpu mask settings
+func RestoreIrqBalanceConfig(irqBalanceConfigFile, irqBannedCPUConfigFile string) error {
```
@cynepco3hahue Thanks for your comments! 
If this happens, can we just rely on `irqSmpAffinityProcFile "/proc/irq/default_smp_affinity"` and restore the config only if this affinity file contains zero value (i.e. system reboot scenario) ?

pperiyasamy replied:
@cynepco3hahue @haircommander have added the logic to restore the config only if default smp affinity proc file has every cpu mask set, this could be a simple approach if none of other applications in the system modified that file beforehand.

cynepco3hahue commented on internal/runtimehandlerhooks/utils.go:
```diff
@@ -121,3 +125,74 @@ func UpdateIRQSmpAffinityMask(cpus, current string, set bool) (cpuMask, bannedCP
+	return !info.IsDir()
+}
+
+func isAllMaskSet(in []byte) bool {
```
Can you please add the comment on the method, that will describe the purpose of the method?

pperiyasamy replied:
of course, done.

cynepco3hahue commented:
/lgtm

haircommander commented:
/retest
/lgtm

pperiyasamy commented:
rebased to latest master.

haircommander commented:
/approve
/lgtm

pperiyasamy commented:
/retest

haircommander commented:
/retest

pperiyasamy commented:
/retest

haircommander commented:
/tetest

pperiyasamy commented:
/retest

saschagrunert commented:
/test integration_fedora

pperiyasamy commented:
/test integration_crun

haircommander commented:
/retest

haircommander commented:
/retest

pperiyasamy commented:
/retest

haircommander commented:
/retest

MarSik commented:
Hi, would it be possible to backport this into 4.7 please? We even have a bug https://bugzilla.redhat.com/show_bug.cgi?id=1934630

haircommander commented:
/cherry-pick release-1.20

let's see if it's clean :)

jianzzha commented:
@haircommander: looks like the above conflict is on completions/zsh/_crio, both edited the line "opts=..." 

Do we need a seperate PR for the backport? I can do that without much effort.

jianzzha commented:
@haircommander: see the above backport PR. The only line changed is in completions/zsh/_crio, line 10
